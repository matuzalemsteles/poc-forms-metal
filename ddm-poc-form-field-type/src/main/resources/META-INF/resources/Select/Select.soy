{namespace Select}

{template .render}
    {@param? _handleClickItem: any}
    {@param? _handleClick: any}
    {@param? editable: bool}
    {@param? helpText: string}
    {@param? id: string}
    {@param? items: list<[
        checked: bool,
        disabled: bool,
        id: string,
        inline: bool,
        label: string,
        name: string,
        showLabel: bool,
        value: string
    ]>}
    {@param? label: string}
    {@param? open: bool}
    {@param? placeholder: string}
    {@param? predefinedValue: string}
    {@param? required: bool}
    {@param? showLabel: bool}
    {@param? spritemap: string}
    {@param? value: string}

    {call FieldBase.render}
        {param contentRenderer kind="html"}
            {call .content}
                {param _handleClickItem: $_handleClickItem /}
                {param _handleClick: $_handleClick /}
                {param editable: $editable /}
                {param id: $id /}
                {param items: $items /}
                {param open: $open /}
                {param placeholder: $placeholder /}
                {param predefinedValue: $predefinedValue /}
                {param spritemap: $spritemap /}
                {param value: $value /}
            {/call}
        {/param}

        {param helpText: $helpText /}
        {param id: $id /}
        {param label: $label /}
        {param required: $required /}
        {param showLabel: $showLabel /}
        {param spritemap: $spritemap /}
    {/call}
{/template}

{template .content}
    {@param? _handleClickItem: any}
    {@param? _handleClick: any}
    {@param? editable: bool}
    {@param? id: string}
    {@param? items: list<[
        checked: bool,
        disabled: bool,
        id: string,
        inline: bool,
        label: string,
        name: string,
        showLabel: bool,
        value: string
    ]>}
    {@param? open: bool}
    {@param? placeholder: string}
    {@param? predefinedValue: string}
    {@param? spritemap: string}
    {@param? value: string}

    <div class="form-builder-select-field input-group-container">
        {let $attributes kind="attributes"}
            class="form-control select-field-trigger"

            {if $id}
                id="{$id}"
            {/if}

            {if not $editable}
                disabled
            {/if}

            data-onclick="{$_handleClick}"
        {/let}

        <div {$attributes}>
            <div class="option-selected option-selected-placeholder">{$value ? $value : ($predefinedValue ? $predefinedValue : $placeholder)}</div>
            <a class="select-arrow-down-container" href="javascript:;">
                {if $spritemap}
                    {call ClayIcon.render}
                        {param spritemap: $spritemap /}
                        {param symbol: 'caret-double' /}
                    {/call}
                {/if}
            </a>
        </div>
    </div>

    {call .listItems}
        {param _handleClickItem: $_handleClickItem /}
        {param items: $items /}
        {param open: $open /}
    {/call}
{/template}

{template .listItems}
    {@param? _handleClickItem: any}
    {@param? items: list<[
        checked: bool,
        disabled: bool,
        id: string,
        inline: bool,
        label: string,
        name: string,
        showLabel: bool,
        value: string
    ]>}

    {@param? open: bool}

    {let $attributes kind="attributes"}
        class="dropdown-menu

        {if $open}
            {sp}dropdown-opened
        {/if}
        "
        data-onclick="{$_handleClickItem}"
        
    {/let}

    {if $items}
        <div {$attributes}>
            <ul class="list-unstyled">
                {foreach $item in $items}
                    <li class="dropdown-item">{$item.value}</li>
                {/foreach}
            </ul>
        </div>
    {/if}
{/template}