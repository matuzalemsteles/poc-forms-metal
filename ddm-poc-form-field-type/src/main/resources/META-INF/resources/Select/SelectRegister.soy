{namespace SelectRegister}

{deltemplate LayoutRenderer.RegisterFieldType variant="'select'"}
    {call Select.render data="all" /}
{/deltemplate}