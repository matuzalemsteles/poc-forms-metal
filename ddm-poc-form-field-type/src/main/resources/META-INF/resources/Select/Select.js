import 'clay-icon';
import {Config} from 'metal-state';
import Component from 'metal-component';
import FieldBase from 'ddm-poc-form-field-type/FieldBase/index.js';
import Soy from 'metal-soy';

import templates from './Select.soy.js';
import SelectRegister from './SelectRegister.soy.js';

class Select extends Component {
    static STATE = {
        /**
         * @default false
         * @instance
         * @memberof Select
         * @type {?bool}
         */
        editable: Config.bool().value(false),

        /**
         * @default undefined
         * @instance
         * @memberof FieldBase
         * @type {?(string|undefined)}
         */
        helpText: Config.string(),

        /**
         * @default undefined
         * @instance
         * @memberof Select
         * @type {?(string|undefined)}
         */        
        id: Config.string(),

        /**
         * @default undefined
         * @instance
         * @memberof Select
         * @type {?array<object>}
         */
        items: Config.arrayOf(Config.shapeOf({
            active: Config.bool().value(false),
            disabled: Config.bool().value(false),
            id: Config.string(),
            inline: Config.bool().value(false),
            label: Config.string(),
            name: Config.string(),
            showLabel: Config.bool().value(true),
            value: Config.string()
        })).value([
            {
                value: 'Option 1'
            },
            {
                value: 'Option 2'
            }
        ]),

        /**
         * @default undefined
         * @instance
         * @memberof Select
         * @type {?(string|undefined)}
         */
        label: Config.string(),

        /**
         * @default undefined
         * @instance
         * @memberof Select
         * @type {?bool}
         */
        open: Config.bool().value(false),

        /**
         * @default Choose an Option
         * @instance
         * @memberof Select
         * @type {?string}
         */
        placeholder: Config.string().value('Choose an Option'),

         /**
         * @default undefined
         * @instance
         * @memberof Select
         * @type {?string}
         */
        predefinedValue: Config.string(),

        /**
         * @default false
         * @instance
         * @memberof Select
         * @type {?bool}
         */
        required: Config.bool().value(false),

        /**
         * @default false
         * @instance
         * @memberof Select
         * @type {?bool}
         */
        showLabel: Config.bool().value(true),

        /**
         * @default undefined
         * @instance
         * @memberof Select
         * @type {?(string|undefined)}
         */
        spritemap: Config.string(),

        /**
         * @default undefined
         * @instance
         * @memberof Select
         * @type {?(string|undefined)}
         */
        value: Config.string(),

        key: Config.string(),
    }

    _handleClickItem(event) {
        const { key } = this;

        this.setState(
            {
                value: event.target.innerText,
                predefinedValue: event.target.innerText
        });

        this.emit('fieldEdit', {
            value: event.target.innerText,
            key,
            originalEvent: event
        });
    }

    _handleClick() {
        this.setState({open : !this.open});
    }
}

Soy.register(Select, templates);

export default Select;