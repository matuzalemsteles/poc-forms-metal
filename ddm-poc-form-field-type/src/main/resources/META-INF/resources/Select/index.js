import Select from './Select.es';

export default Select;