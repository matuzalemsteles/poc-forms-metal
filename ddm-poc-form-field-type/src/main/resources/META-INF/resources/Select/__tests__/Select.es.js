import Select from '../Select.es';
import { dom as MetalTestUtil} from 'metal-dom';

let component;
let spritemap = 'icons.svg';

describe('Select', () => {
    afterEach(() => {
        if (component) {
			component.dispose();
        }
    });

    it('should be not edidable', () => {
        component = new Select({
            editable: false,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a helptext', () => {
        component = new Select({
            helpText: 'Type something',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have an id', () => {
        component = new Select({
            id: 'ID',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should render items', () => {
        component = new Select({
            items: [{
                checked: false,
                disabled: false,
                id: 'id',
                inline: false,
                label: 'label',
                name: 'name',
                showLabel: true,
                value: 'item'
            }]
        });

        expect(component).toMatchSnapshot();
    });

    it('should render no items when items come empty', () => {
        component = new Select({
            items: []
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a label', () => {
        component = new Select({
            label: 'label',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should be closed by default', () => {
        component = new Select({
            open: false,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have class dropdown-opened when it\'s opened', () => {
        component = new Select({
            open: true,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a placeholder', () => {
        component = new Select({
            placeholder: 'Placeholder',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a predefinedValue', () => {
        component = new Select({
            predefinedValue: 'Select',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should not be required', () => {
        component = new Select({
            required: false,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should render Label if showLabel is true', () => {
        component = new Select({
            label: 'text',
            showLabel: true,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    
    it('should have a spritemap', () => {
        component = new Select({
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a value', () => {
        component = new Select({
            value: 'value',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a key', () => {
        component = new Select({
            key: 'key',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should emit a field edit event when an item is selected', () => {
        const handleFieldEdit = jest.fn();
        const events = {fieldEdit: handleFieldEdit};
        component = new Select({
            items: [
                {
                    checked: false,
                    disabled: false,
                    id: 'id',
                    inline: false,
                    label: 'label',
                    name: 'name',
                    showLabel: true,
                    value: 'item'
                }
            ],
            spritemap: spritemap,
            events
        });

        MetalTestUtil.triggerEvent(component.element.querySelector('.dropdown-menu'), 'click', {});

        expect(handleFieldEdit).toHaveBeenCalled();
    });

    it('should open dropdown when select is clicked', () => {
        component = new Select({
            items: [
                {
                    checked: false,
                    disabled: false,
                    id: 'id',
                    inline: false,
                    label: 'label',
                    name: 'name',
                    showLabel: true,
                    value: 'item'
                }
            ],
            spritemap: spritemap
        });

        MetalTestUtil.triggerEvent(component.element.querySelector('.select-field-trigger'), 'click', {});
        expect(component.getState().open).toBe(true);
    });

    it('should propagate the field edit event', () => {
        component = new Select({
            items: [
                {
                    checked: false,
                    disabled: false,
                    id: 'id',
                    inline: false,
                    label: 'label',
                    name: 'name',
                    showLabel: true,
                    value: 'item'
                }
            ],
            spritemap: spritemap
        });

        const spy = jest.spyOn(component, 'emit');

        MetalTestUtil.triggerEvent(component.element.querySelector('.dropdown-menu'), 'click', {});

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
			'fieldEdit',
			expect.any(Object)
		);
    });
});
