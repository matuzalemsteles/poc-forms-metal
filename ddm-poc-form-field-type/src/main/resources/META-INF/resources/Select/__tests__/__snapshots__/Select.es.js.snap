// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should be closed by default 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should be not edidable 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have a helptext 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text">Type something</span>
</div>
`;

exports[`Select should have a key 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have a label 1`] = `
<div class="form-group">
  <label>label </label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have a placeholder 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Placeholder</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have a predefinedValue 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Select</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have a spritemap 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have a value 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">value</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have an id 1`] = `
<div class="form-group">
  <label for="ID"></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" id="ID" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should have class dropdown-opened when it's opened 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu dropdown-opened">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should not be required 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should render Label if showLabel is true 1`] = `
<div class="form-group">
  <label>text </label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-double" focusable="false">
          <title>caret-double</title>
          <use xlink:href="icons.svg#caret-double"></use>
        </svg>
      </a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">Option 1</li>
      <li class="dropdown-item">Option 2</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should render items 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;"></a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled">
      <li class="dropdown-item">item</li>
    </ul>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Select should render no items when items come empty 1`] = `
<div class="form-group">
  <label></label>
  <div class="form-builder-select-field input-group-container">
    <div class="form-control select-field-trigger" disabled="">
      <div class="option-selected option-selected-placeholder">Choose an Option</div>
      <a class="select-arrow-down-container" href="javascript:;"></a>
    </div>
  </div>
  <div class="dropdown-menu">
    <ul class="list-unstyled"></ul>
  </div>
  <span class="form-text"></span>
</div>
`;
